File: OJ practices/LeetCode题目集/30. Substring with Concatenation of All Words(hard).py
'''
@Author: SourDumplings
@Date: 2020-01-29 12:50:27
@Link: https://github.com/SourDumplings/
@Email: changzheng300@foxmail.com
@Description: https://leetcode.com/problems/substring-with-concatenation-of-all-words/
'''
from typing import List


class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        res = []
        if len(words) == 0:
            return res

        lOfWord = len(words[0])
        lOfSubstring = 0
        l = len(s)

        d = dict()
        for word in words:
            d[word] = d.get(word, 0) + 1
            lOfSubstring += lOfWord

        for i in range(0, l - lOfSubstring + 1):
            flag = True
            temp = d.copy()
            subStr = s[i:i + lOfSubstring]
            for j in range(0, lOfSubstring - lOfWord + 1, lOfWord):
                word = subStr[j:j + lOfWord]
                num = temp.get(word)
                if num is None or num == 0:
                    flag = False
                    break
                else:
                    temp[word] -= 1
            if flag:
                res.append(i)

        return res

File: OJ practices/LeetCode题目集/30. Substring with Concatenation of All Words(hard).java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/*
 * @Author: SourDumplings
 * @Date: 2020-01-29 12:08:58
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/substring-with-concatenation-of-all-words/
 */

class Solution
{
    public List<Integer> findSubstring(String s, String[] words)
    {
        List<Integer> res = new ArrayList<>();
        if (words.length == 0)
        {
            return res;
        }

        Map<String, Integer> map = new HashMap<>();
        int lOfWord = words[0].length();
        int lOfSubString = 0;
        for (String word : words)
        {
            map.put(word, map.getOrDefault(word, 0) + 1);
            lOfSubString += lOfWord;
        }
        int l = s.length();
        for (int i = 0; i <= l - lOfSubString; i++)
        {
            String subString = s.substring(i, i + lOfSubString);
            Map<String, Integer> temp = (Map<String, Integer>) ((HashMap) map).clone();
            boolean flag = true;
            for (int j = 0; j <= lOfSubString - lOfWord; j += lOfWord)
            {
                String word = subString.substring(j, j + lOfWord);
                if (!temp.containsKey(word) || temp.get(word) == 0)
                {
                    flag = false;
                    break;
                }
                else
                {
                    temp.put(word, temp.get(word) - 1);
                }
            }
            if (flag)
            {
                res.add(i);
            }
        }
        return res;
    }
}